src/ServiceControl/CompositeViews/Messages/MessageViewQueryExtensions.cs

```diff
@@ -0,0 +1,134 @@
+    using System.IO;
+    using System.Linq;
+    using System.Net.Http;
+    using System.Text;
+    using System.Threading.Tasks;
+    using Nancy;
+    using Newtonsoft.Json;
+    using NServiceBus.Logging;
+    using ServiceBus.Management.Infrastructure.Extensions;
+    using ServiceBus.Management.Infrastructure.Nancy;
+    using ServiceBus.Management.Infrastructure.Nancy.Modules;
+    using HttpStatusCode = System.Net.HttpStatusCode;
+
+    public static class MessageViewQueryExtensions
+    {
+        static JsonSerializer jsonSerializer = JsonSerializer.Create(JsonNetSerializer.CreateDefault());
+        static ILog logger = LogManager.GetLogger(typeof(MessageViewQueryExtensions));
+
+        private static QueryResult EmptyResult = new QueryResult(new List<MessagesView>(), new QueryStatsInfo(string.Empty, DateTime.MinValue, 0, 0));
+
+        public static async Task<dynamic> CombineWithRemoteResults(this BaseModule module, QueryResult localQueryResult)
+        {
+            var httpClientFactory = module.HttpClientFactory; // for testing purposes
+            var remotes = module.Settings.RemoteInstances;
+            var currentRequest = module.Request;
+            var negotiator = module.Negotiate;
+
+            if (remotes.Length > 0)
+            {
+                await UpdateLocalQueryResultWithRemoteData(currentRequest, httpClientFactory, remotes, localQueryResult).ConfigureAwait(false);
+
+                localQueryResult.Messages.Sort(MessageViewComparer.FromRequest(currentRequest));
+            }
+
+            return negotiator.WithPartialQueryResult(localQueryResult, currentRequest);
+        }
+
+        static async Task UpdateLocalQueryResultWithRemoteData(Request currentRequest, Func<HttpClient> httpClientFactory, string[] remotes, QueryResult localQueryResult)
+        {
+            var tasks = new List<Task<QueryResult>>(remotes.Length);
+            // ReSharper disable once LoopCanBeConvertedToQuery
+            foreach (var remote in remotes)
+            {
+                tasks.Add(FetchAndParse(currentRequest, httpClientFactory, remote));
+            }
+
+            var highestTotalCount = localQueryResult.QueryStats.TotalCount;
+            var totalCount = localQueryResult.QueryStats.TotalCount;
+            var lastModified = DateTime.MinValue;
+            var etagBuilder = new StringBuilder();
+            foreach (var queryResult in await Task.WhenAll(tasks))
+            {
+                localQueryResult.Messages.AddRange(queryResult.Messages);
+
+                var header = queryResult.QueryStats;
+                totalCount += header.TotalCount;
+
+                if (header.HighestTotalCountOfAllTheInstances > highestTotalCount)
+                {
+                    highestTotalCount = header.HighestTotalCountOfAllTheInstances;
+                }
+
+                if (header.LastModified > lastModified)
+                {
+                    lastModified = header.LastModified;
```

Comment:
There's some weirdness here when there's clock drift between Master and Slave.

- I query. Master locally has a LastMod value of t5. Slave reports a LastMod of t3
- My response has a LastMod of t5 (the max of t5 and t3)
- I query again but this time I pass up my LastModified header (t5). Master locally reports t5 (unchanged), Slave reports a LastMod of t4 (it changed, but in the past relative to Master)
- My response comes back as NotModified but that isn't true _even though my etags don't match_.
- If you shut ingestion down on both instances, a good caching client will _never_ see the updated values

Unfortunately, taking the MIN doesn't work either. If you take the MIN then shutting ingestion off on any instance will freeze the result set.

As long as _any_ instance is updating, taking the MAX will work out over time.

Reply:
In short, I think this is OK but it has a limitation that we need to be aware of. If we think it's likely that we'll see cases where NO instance is ingesting new messages then we might want to drop the LastModified header as it might be misleading.

Reply:
It's also probably worth noting that the SI client only sends If-None-Match (etag), and does not send If-Modified-Since (last modified). For our use case this isn't a big deal to get right the first time.

Reply:
So what you are saying we can skip the last modified entirely in the aggregated mode?